[dynamo][eval_frame] Unify cache entry and frame_state on the same co_extra index

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107388
* #107358
* #107117
* __->__ #106917

Handling follow up from https://github.com/pytorch/pytorch/pull/106413#discussion_r1288971923


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

anijain2305 committed:
[dynamo][eval_frame] Unify cache entry and frame_state on the same co_extra index

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][eval_frame] Unify cache entry and frame_state on the same co_extra index"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][eval_frame] Unify cache entry and frame_state on the same co_extra index"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -322,26 +336,57 @@ static void destroy_cache_entry(CacheEntry* e) {
-  _PyCode_GetExtra((PyObject*)code, cache_entry_extra_index, (void*)&extra);
+static ExtraState* create_extra_state(
+    CacheEntry* cache_entry,
+    PyObject* frame_state) {
```
Document ownership contract on arguments and outputs

ezyang replied:
(steals)

anijain2305 committed:
Update on "[dynamo][eval_frame] Unify cache entry and frame_state on the same co_extra index"


Handling follow up from https://github.com/pytorch/pytorch/pull/106413#discussion_r1288971923


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][eval_frame] Unify cache entry and frame_state on the same co_extra index"


Handling follow up from https://github.com/pytorch/pytorch/pull/106413#discussion_r1288971923


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -745,17 +798,17 @@ static PyObject* _custom_eval_frame(
-    destroy_cache_entry(extra);
-    set_cache_entry(frame->f_code, SKIP_CODE);
+    destroy_extra_state(frame->f_code);
+    set_extra_state(frame->f_code, SKIP_CODE);
```
The destroy above shouldn't be necessary, right? I would expect `set_extra_state` to care of deallocating the previous entry.

anijain2305 replied:
`set_extra_state` is really simple, it just hides a private API - `_PyCode_SetExtra`

```
inline static void set_extra_state(PyCodeObject* code, ExtraState* extra_state) {
  // Sets the extra state on the extra scrach space of the code object.
  _PyCode_SetExtra((PyObject*)code, extra_index, extra_state);
}
```

I don't want set_extra_state` to do deallocation. One could just alter the items in the linked list and then set the extra_state. In that case, we would not want to deallocate, we would just want to override. So, having different functions handle the deallocation and setting the extra field seems ok to me.

ezyang replied:
OK, then you should document that set_extra_state does NOT deallocate the previous entry and it's the responsibility of the caller to dispose of the previous entry.

anijain2305 replied:
I went back and made `set_extra_state` to deallocate the prev entry.

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -745,17 +798,17 @@ static PyObject* _custom_eval_frame(
   } else if (result != Py_None) {
     DEBUG_TRACE("create cache %s", get_frame_name(frame));
-    extra = create_cache_entry(extra, result);
+    CacheEntry* new_cache_entry = create_cache_entry(cache_entry, result);
```
But... you don't end up using cache_entry? Overwriting the old cache_entry is stylistically cleaner because if create_cache_entry steals ownership you're technically not allowed to access the old cache entry anymore right.

anijain2305 replied:
Good point. Making the change.

ezyang replied:
You haven't actually done this change yet

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -724,16 +775,18 @@ static PyObject* _custom_eval_frame(
   if (frame_state == NULL) {
     // TODO(voz): Replace this dict with a real FrameState object.
     frame_state = PyDict_New();
-    set_frame_state(frame->f_code, frame_state);
```
I think you leak `frame_state` now when you create a skip.

anijain2305 replied:
Very nice catch!

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -299,6 +302,17 @@ typedef struct cache_entry {
 
+// An abstraction to encapsulate all the information to be saved on the extra
+// scratch space of the code object. This enables us to reuse the same co_extra
+// field, saving extra allocations.
```
Well, you're not actually saving any allocations at the moment, since you needed to malloc ExtraState so you net out two allocations (one for the extra scratch, one for ExtraState) whereas before you had two for extra scratch.

anijain2305 replied:
I will modify the comment. The goal is to not have two allocations for extra scratch. @jansel mentioned that its very expensive to have more allocations for extra. I don't understand why yet. Will try to dig deeper to understand this point.

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -438,10 +484,14 @@ static PyObject* lookup(CacheEntry* e, THP_EVAL_API_FRAME_OBJECT *frame, CacheEn
+        e->next = old_cache_entry;
+
+        // Update the extra state to reflect the updated cache
+        extra->cache_entry = e;
```
What if extra is NULL?

anijain2305 replied:
It can't be NULL here. CacheEntry* e is not NULL, otherwise we would have returned at the beginning of this function. Since, CacheEntry is extracted from the extra, it can't be NULL here.

anijain2305 committed:
Update on "[dynamo][eval_frame] Unify cache entry and frame_state on the same co_extra index"


Handling follow up from https://github.com/pytorch/pytorch/pull/106413#discussion_r1288971923


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][eval_frame] Unify cache entry and frame_state on the same co_extra index"


Handling follow up from https://github.com/pytorch/pytorch/pull/106413#discussion_r1288971923


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][eval_frame] Unify cache entry and frame_state on the same co_extra index"


Handling follow up from https://github.com/pytorch/pytorch/pull/106413#discussion_r1288971923


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

jansel commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -915,17 +1065,10 @@ static struct PyModuleDef _module = {
-
-  dynamic_frame_state_extra_index = _PyEval_RequestCodeExtraIndex(ignored);
-  if (dynamic_frame_state_extra_index < 0) {
+  extra_index = _PyEval_RequestCodeExtraIndex(ignored);
```
A pre-existing issue, but we should be passing a valid `freefunc` here rather than `ignored`.  If the code object is deallocated (I suppose pretty uncommon as it only comes from eval(),etc) we will leak this.

anijain2305 replied:
This is little more complicated, so I would prefer to handle it in different PR. `freefunc` is called everytime we call `SetExtra`. So, we will have to move things around little bit to ensure 

1) we don't double destroy extra states
2) make a copy of extra instead of overwriting to do MRU optimization

I will do a followup in a separate PR.

![image](https://github.com/pytorch/pytorch/assets/13822661/e33596ee-02e0-45f0-b3a4-7a35608f56ee)

anijain2305 replied:
Follow up in this PR - https://github.com/pytorch/pytorch/pull/107117

After working on @ezyang's comments, I found it easy to set the deleter here. Keeping it separate for easy reviews. But, happy to merge it in 1 PR as well.

anijain2305 committed:
Update on "[dynamo][eval_frame] Unify cache entry and frame_state on the same co_extra index"


Handling follow up from https://github.com/pytorch/pytorch/pull/106413#discussion_r1288971923


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][eval_frame] Unify cache entry and frame_state on the same co_extra index"


Handling follow up from https://github.com/pytorch/pytorch/pull/106413#discussion_r1288971923


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -299,6 +305,20 @@ typedef struct cache_entry {
+  CacheEntry* cache_entry;
+  // Frame state to detect dynamic shape dims
+  FrameState* frame_state;
+} ExtraState;
```
I wonder if you do want to turn this into an honest to goodness PyObject eventually, as you did in your other PR.

anijain2305 replied:
I am not planning to. It was needed because I was using PyDict earlier and i needed PyObject equivalent of CacheEntry then. Atleast for this PR, I dont need it.

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -854,7 +974,7 @@ static PyObject* skip_code(PyObject* dummy, PyObject* obj) {
     return NULL;
   }
-  set_cache_entry((PyCodeObject*)obj, SKIP_CODE);
+  set_extra_state((PyCodeObject*)obj, SKIP_CODE);
```
This seems to be a pre-existing problem, but you need to check if the original entry was non-NULL and deallocate it.

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -322,12 +342,99 @@ static void destroy_cache_entry(CacheEntry* e) {
+}
+
+
+inline static void create_and_set_extra_state(
```
invariant: there must not be any pre-existing extra state on code

anijain2305 replied:
Added

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -322,12 +342,99 @@ static void destroy_cache_entry(CacheEntry* e) {
+  set_extra_state(code, extra_state);
+}
+
+inline static void override_cache_entry_on_extra(
```
invariant: there must not be any pre-existing state on extra

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -745,17 +858,28 @@ static PyObject* _custom_eval_frame(
+      create_and_set_extra_state(frame->f_code, new_cache_entry, frame_state);
+    } else {
+      // Update the existing cache_entry on the extra scratch space.
+      override_cache_entry_on_extra(frame->f_code, extra, new_cache_entry);
```
This seems extremely suspect. override_cache_entry_on_extra sets cache entry. But extra is already set to your malloc field.

The problem is that when you consolidated the extra and frame state fields, you cargo culted the setter logic. But you do not need to set the NEW extra entry anymore, because modifying the cache entry field on the struct is sufficient!

anijain2305 replied:
Yes, you are right. I realized this well that I am doing something very dumb.

ezyang reviewed (changes requested):
I'm stopping review again because auditing C code by hand is very time consuming and I'm still finding things that are very suspect, if not outright bugs.

Here is your homework: For every function in the file, document (1) the ownership contract for input/output (does it borrow or steal args? Does it return an owning ref or a borrowed ref)? Do it even for obvious functions. Then (2) audit every call site for every function and ensure that it is abiding by the contract in question. (3) Do the same for functions that have any other preconditions (set_extra_state, I'm looking at you); document the preconditions, and then manually verify the preconditions are upheld.)

anijain2305 committed:
Update on "[dynamo][eval_frame] Unify cache entry and frame_state on the same co_extra index"


Handling follow up from https://github.com/pytorch/pytorch/pull/106413#discussion_r1288971923


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

anijain2305 commented:
@ezyang Thanks for spending time on the review. I documented the ownership rules. Added CHECK for invariants wherever possible. I am new to reference ownership, so I am not very confident. The main change here is that `set_extra_state` becomes owner and also destroys the pre-existing object on the co_extra, before setting the new extra state. This has simplified the code at bunch of places.

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -299,9 +305,24 @@ typedef struct cache_entry {
 static CacheEntry* create_cache_entry(
     CacheEntry* next,
     PyObject* guarded_code) {
+  // Onwership contract - Returns a new reference
```
Ownership

ezyang replied:
This steals next, but only borrows `guarded_code`

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -322,12 +343,150 @@ static void destroy_cache_entry(CacheEntry* e) {
+  // args
+  //  - extra_state: Borrowed
+  // return
+  //  - CacheEntry: Borrowed. set_extra_state is the owner.
```
"set_extra_state is the owner" this doesn't make any sense

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -322,12 +343,150 @@ static void destroy_cache_entry(CacheEntry* e) {
   return extra;
 }
 
+inline static void destroy_extra_state(PyCodeObject* code) {
```
Borrows code

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -322,12 +343,150 @@ static void destroy_cache_entry(CacheEntry* e) {
+  }
+}
+
+inline static void set_extra_state(PyCodeObject* code, ExtraState* extra_state) {
```
code is borrowed

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -322,12 +343,150 @@ static void destroy_cache_entry(CacheEntry* e) {
+  //  - there is no return, but the extra_state is stolen, so it becomes
+  //  set_extra_state responsibility to clean it up. It will be deleled during
+  //  the reset_code/skip, when the set_extra_state is called with
+  //  NULL/SKIP_CODE.
```
There is another invariant, which is you must not try to set extra state for an extra state that is ALREADY on that code object. You will destroy the extra state and then write something invalid into SetExtra. We could test this invariant by running `PyCode_GetExtra` first, if you don't mind the extra CPython API call (it might be fine, I'm not sure.)

anijain2305 replied:
Added. Its fine, its not in the critical path. This code will get triggered only when we are recompiling.

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -322,12 +343,150 @@ static void destroy_cache_entry(CacheEntry* e) {
+  // Ownership contract
+  // args
+  //  - cache_entry: Borrowed
+  //  - frame_state: Borrowed
```
This contradicts the text you wrote above

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -322,12 +343,150 @@ static void destroy_cache_entry(CacheEntry* e) {
+
+
+inline static void create_and_set_extra_state(
+    PyCodeObject* code,
```
Code is borrowed

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -362,20 +522,23 @@ PyObject* _debug_get_cache_entry_list(PyObject* self, PyObject* args) {
-  _PyCode_GetExtra((PyObject*)code, dynamic_frame_state_extra_index, (void*)&extra);
-  return extra;
-}
+static inline PyObject* call_callback(
```
Is this code motion only?

anijain2305 replied:
Yes

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -322,12 +343,150 @@ static void destroy_cache_entry(CacheEntry* e) {
+
+inline static void override_cache_entry_on_extra(
+    ExtraState* extra,
+    CacheEntry* new_cache_entry) {
```
`new_cache_entry` is stolen

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -438,10 +601,14 @@ static PyObject* lookup(CacheEntry* e, THP_EVAL_API_FRAME_OBJECT *frame, CacheEn
         prev->next = e->next;
-        e->next = extra;
-        set_cache_entry(frame->f_code, e);
+        e->next = old_cache_entry;
```
Yup, here it is. This violates the contracts: you can't set old_cache_entry to an owning field, because you only got a borrowed reference from extract_cache_entry

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -663,11 +830,18 @@ static PyObject* _custom_eval_frame(
+  if (frame_state == NULL) {
+    // frame_state becomes a new reference. It gets cleaned up by
+    // set_extra_state or skip later.
+    frame_state = PyDict_New();
```
Ugh. Why not just immediately set it into extra? As written here, the variable `frame_state` is either owning or non-owning, depending on whether or not there was a frame state on extra. That way lies madness. Just populate it immediately on access.

anijain2305 replied:
Introduced a new function - init_and_set_extra_state which inits the extra_state with (CacheEntry=NULL, FrameState=PyDict). This resolves the owning or non-owning depending on the code path.

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -745,17 +912,27 @@ static PyObject* _custom_eval_frame(
+      create_and_set_extra_state(frame->f_code, new_cache_entry, frame_state);
+    } else {
+      // Update the existing cache_entry on the extra scratch space.
+      override_cache_entry_on_extra(extra, new_cache_entry);
```
Yeah this has the same problem as the other call site. There's one more problem, which is you're using extra == NULL to demarcate whether or not there was already a cache entry on extra, but that's actually an extra invariant. Couldn't you imagine a case where you have no cache entries, but you DO have a frame state? It seems better to just get the truth from the horse's mouth.

anijain2305 replied:
Thanks for laying it out for me. Yeah, I inlined the code. There is still violation but it is contained in a local region, so hopefully its acceptable?

ezyang reviewed (changes requested):
Still some organizational difficulties

anijain2305 committed:
Update on "[dynamo][eval_frame] Unify cache entry and frame_state on the same co_extra index"


Handling follow up from https://github.com/pytorch/pytorch/pull/106413#discussion_r1288971923


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][eval_frame] Unify cache entry and frame_state on the same co_extra index"


Handling follow up from https://github.com/pytorch/pytorch/pull/106413#discussion_r1288971923


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

anijain2305 commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -322,12 +343,150 @@ static void destroy_cache_entry(CacheEntry* e) {
+
+  // Invariant - Extra state should have been set before, therefore it should not be NULL.
+  NULL_CHECK(extra);
+  extra->cache_entry = new_cache_entry;
```
Directly inlined the code.

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -322,12 +348,126 @@ static void destroy_cache_entry(CacheEntry* e) {
+  // args
+  //  - code: Borrowed
+  // return
+  //  - extra_state: Borrowed. set_extra_state is the owner.
```
I still don't know what it means for a function to "own" data

anijain2305 replied:
I will just remove that line. What I really meant here was the set_extra_State is responsible for cleaning it up, but it does not make sense here.

anijain2305 committed:
Update on "[dynamo][eval_frame] Unify cache entry and frame_state on the same co_extra index"


Handling follow up from https://github.com/pytorch/pytorch/pull/106413#discussion_r1288971923


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][eval_frame] Unify cache entry and frame_state on the same co_extra index"


Handling follow up from https://github.com/pytorch/pytorch/pull/106413#discussion_r1288971923


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -322,12 +348,126 @@ static void destroy_cache_entry(CacheEntry* e) {
+  //  - extra_state: Stolen
+  // return
+  //  - there is no return, but the extra_state is stolen, so it becomes
+  //  set_extra_state responsibility to clean it up. It will be deleled during
```
deleted

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -322,12 +348,126 @@ static void destroy_cache_entry(CacheEntry* e) {
+}
+
+inline static void set_extra_state(PyCodeObject* code, ExtraState* extra_state) {
+  // Clears the existing object sitting on the extra scratch spance and sets it
```
space

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -438,10 +582,14 @@ static PyObject* lookup(CacheEntry* e, THP_EVAL_API_FRAME_OBJECT *frame, CacheEn
         prev->next = e->next;
-        e->next = extra;
-        set_cache_entry(frame->f_code, e);
+        e->next = old_cache_entry;
```
This is still a violation of borrowing semantics. You're not allowed to steal something you got from a borrow (`extract_cache_entry`). Probably good enough to just say `extra->cache_entry` instead.

ezyang commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -745,17 +893,20 @@ static PyObject* _custom_eval_frame(
   } else if (result != Py_None) {
     DEBUG_TRACE("create cache %s", get_frame_name(frame));
-    extra = create_cache_entry(extra, result);
+    CacheEntry* new_cache_entry = create_cache_entry(cache_entry, result);
```
This is also technically a violation of borrow semantics, as `cache_entry` came from a borrow. Suggest `extra->cache_entry` instead.

anijain2305 committed:
Update on "[dynamo][eval_frame] Unify cache entry and frame_state on the same co_extra index"


Handling follow up from https://github.com/pytorch/pytorch/pull/106413#discussion_r1288971923


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]
